lib/metadata/chordMetadata.cpp

```diff
@@ -41,3 +43,123 @@ chordMetadata::chordMetadata() :
         time_downsampling_fpga[f] = -1;
     }
 }
+
+struct chordMetadataFormat {
+    int32_t max_dim;
+    int32_t max_dimname;
+    int32_t max_freq;
+
+    int32_t frame_counter;
+
+    // chordDataType type;
+    int32_t type;
+
+    int32_t dims;
+    int32_t dim[CHORD_META_MAX_DIM];
+    char dim_name[CHORD_META_MAX_DIM][CHORD_META_MAX_DIMNAME]; // "F", "Tbar", "D", etc
+    int64_t strides[CHORD_META_MAX_DIM];
+    int64_t offset;
+
+    // One-hot arrays?
+    int32_t n_one_hot;
+    char onehot_name[CHORD_META_MAX_DIM][CHORD_META_MAX_DIMNAME];
+    int32_t onehot_index[CHORD_META_MAX_DIM];
+
+    // Per-frequency arrays
+    int32_t nfreq;
+
+    // frequencies -- integer (0-2047) identifier for FPGA coarse frequencies
+    int32_t coarse_freq[CHORD_META_MAX_FREQ];
+
+    // the upchannelization factor that each frequency has gone through (1 for = FPGA)
+    int32_t freq_upchan_factor[CHORD_META_MAX_FREQ];
+
+    // Time sampling -- for each coarse frequency channel, 2x the FPGA
+    // sample number of the first sample.  The 2x is there to handle
+    // the upchannelization case, where 2 or more samples may get
+    // averaged, producing a new sample that is effectively halfway in
+    // between them, ie, at a half-FPGAsample time.
+    int64_t half_fpga_sample0[CHORD_META_MAX_FREQ];
+
+    // Time sampling -- for each coarse frequency channel, the factor
+    // by which the time samples have been downsampled relative to
+    // FPGA samples.
+    int32_t time_downsampling_fpga[CHORD_META_MAX_FREQ];
+};
+
+size_t chordMetadata::get_serialized_size() {
+    return chimeMetadata::get_serialized_size() + sizeof(chordMetadataFormat);
+}
+
+size_t chordMetadata::set_from_bytes(const char* bytes, size_t length) {
```

Comment:
Are these functions tested?

I don't think these functions work well. They implicitly depend on global `#define` e.g. for the number of frequencies, and there are not checks for these, and the case where they change isn't handled. I would simply rip out this kind of binary serialization, it's worse than common blocks in Fortran. There are nice libraries (e.g. https://uscilab.github.io/cereal/) that we could use instead if we needed. Otherwise I strongly prefer serializing to YAML, e.g. via ASDF.

Reply:
I completely agree, we should serialize to a better format (this class has JSON serialization support).  But the CHIME code uses binary serialization extensively, so for backward compatibility I kept that.  We could just have the chordMetadata refuse to serialize itself into binary if you prefer!